README.md
# grosso

## Grosso worm game

Grosso was designed for Unix systems and for 80x24 character terminal. It also runs on Cygwin.

![alt text](https://github.com/tommijoutsiniemi/grosso/blob/main/misc/grosso-ss.png?raw=true)

## Build
	make all

### Dependencies
Grosso uses ncurses, so something like libncurses-dev is needed.

## Install
	sudo make install

## Instructions for playing:

![alt text](https://github.com/tommijoutsiniemi/grosso/blob/main/misc/grosso-help.png?raw=true)


ChangeLog
1997-10-14  Tommi Joutsiniemi  <Tommi.Joutsiniemi@nic.fi>

	* Debianized: Added support for DESTDIR installation prefix and
	changed the locations, owners and modes of all installed files.

	* Reorganized the source tree.

1997-10-11  Tommi Joutsiniemi  <Tommi.Joutsiniemi@nic.fi>

	* Added color support.

Tue Mar 25 23:03:43 1997  Tommi Joutsiniemi  <Tommi.Joutsiniemi@nic.fi>

	* Moved make definitions to a separate file, and created two
 	versions of it: make.defs.como and make.defs.g++.

	* Modified to compile with Comeau C++, mostly by dropping streams
 	and fixing includes.

Sun Mar 23 12:57:18 1997  Tommi Joutsiniemi  <Tommi.Joutsiniemi@nic.fi>

	* Field.cc (Field): Removed class Yard, merged to class Field.

Sat Mar 22 12:55:57 1997  Tommi Joutsiniemi  <Tommi.Joutsiniemi@nic.fi>

	* Field.hh: Moved Field's typedefs inside the class. Now much more
 	easier and straightforward to use.

Thu Oct  3 22:24:42 1996  Tommi Joutsiniemi  <Tommi.Joutsiniemi@nic.fi>

	* First functional version, no known bugs!


src/Fred.cc
//
// Grosso: Fred.cc
//

// 2002/06/09 10:29:50 tommi

///////////////////////////////////////////////////////////////////////////

#ifndef MAKEDEPEND
#include <stdlib.h>
#endif

#include "config.hh"
#include "Fred.hh"

///////////////////////////////////////////////////////////////////////////

Fred::Fred(Field* field, Field::Item head, Field::ColourNum fg, Field::ColourNum bg, int probab) :
  Worm(field, head, fg, bg), probab_(probab)
{
}

///////////////////////////////////////////////////////////////////////////

void
Fred::Reset(Field::Dim headx, Field::Dim heady, int dir, int length)
{
  Worm::Reset(headx, heady, dir, length);
  strength_ = 100;
  dead_ = false;
}

///////////////////////////////////////////////////////////////////////////

Fred::MoveResult
Fred::TryTurn()
{
  // try a random turn
  int turn = rand()&1 ? LEFT : RIGHT;
  if (IsGoable(Peek(turn))) {
    Turn(turn);
  }
  else if (IsGoable(Peek(-turn))) {
    Turn(-turn);
  }
  else {
    // can't turn, give up
    return GOT_BLOCKED;
  }

  // turned if here, so lengthen and actually take the turn
  Lengthen(10);
  return Move();
}

///////////////////////////////////////////////////////////////////////////

Fred::MoveResult
Fred::TryMove(int)
{
  MoveResult result;

  if (dead_) {
    return IS_DEAD;
  }

  result = GOT_BLOCKED;
  // try to move, sometimes try to turn first
  if ((rand()%1000) < probab_) {
    // first try turning, if still blocked try forward
    result = TryTurn();
    if (result == GOT_BLOCKED && IsGoable(Peek(FORWARD))) {
      result = Move();
    }
  }
  else {
    // first try forward, if still blocked try turning
    if (IsGoable(Peek(FORWARD))) {
      result = Move();
    }
    if (result == GOT_BLOCKED) {
      result = TryTurn();
    }
  }

  if (result == GOT_BLOCKED) {
    if (!MoveTail()) {
      // couldn't move the head nor the tail
      strength_--;
      if (strength_ == 0) {
        dead_ = true;
        result = DIED;
        ShowHead(Field::EMPTY);
      }
    }
    else {
      // got blocked, but shrinked successfully, so "moved"
      result = MOVED;
    }
  }

  return result;
}


src/Worm.cc
//
// Grosso: Worm.cc
//

// 2002/06/09 10:29:49 tommi

///////////////////////////////////////////////////////////////////////////

#include "config.hh"
#include "Field.hh"
#include "Worm.hh"

///////////////////////////////////////////////////////////////////////////

static const Field::Dim xdimd[] = {  0,  1,  0, -1};
static const Field::Dim ydimd[] = { -1,  0,  1,  0};
static const Field::Item tail[] = { 'A', '>', 'V', '<'};

///////////////////////////////////////////////////////////////////////////

Worm::Worm(Field* field, Field::Item head, Field::ColourNum fg, Field::ColourNum bg)
{
  field_ = field;
  head_ = head;
  lengthen_ = 10;
  dir_ = 2;
  headx_ = 0;
  heady_ = 0;
  tailx_ = 0;
  taily_ = 0;
  colourNum_ = field_->NewColour(fg, bg);
}

///////////////////////////////////////////////////////////////////////////

void
Worm::Reset(Field::Dim headx, Field::Dim heady, int dir, int length)
{
  lengthen_ = length;
  dir_ = Direction(dir);
  headx_ = headx;
  heady_ = heady;
  tailx_ = headx;
  taily_ = heady;

  field_->PutItem(headx_, heady_, tail[dir_], colourNum_);
}

///////////////////////////////////////////////////////////////////////////

Field::Item
Worm::Peek(int turn)
{
  int dir = Direction(dir_+turn);
  Field::Dim x = headx_;
  Field::Dim y = heady_;

  field_->Move(x, y, xdimd[dir], ydimd[dir]);
  return field_->GetItem(x, y);
}

///////////////////////////////////////////////////////////////////////////

Worm::MoveResult
Worm::MoveHead()
{
  Field::Item i;

  field_->PutItem(headx_, heady_, tail[dir_], colourNum_);
  field_->Move(headx_, heady_, xdimd[dir_], ydimd[dir_]);
  i = field_->GetItem(headx_, heady_);
  field_->PutItem(headx_, heady_, head_, colourNum_);

  switch (i) {
  case FOOD:
    Lengthen(10);
    return ATE;
  case Field::EMPTY:
    return MOVED;
  default:
    return CRASHED;
  }
}

///////////////////////////////////////////////////////////////////////////

bool
Worm::MoveTail()
{
  int dir;
  Field::Item i;
  bool result;

  if (lengthen_ > 0) {
    lengthen_--;
    result = true;
  }
  else {
    i = field_->GetItem(tailx_, taily_);
    result = (i != head_);
    for (dir = 0; dir < NUM_DIRECTIONS; dir++) {
      if (i == tail[dir]) {
        field_->PutItem(tailx_, taily_, Field::EMPTY, colourNum_);
        field_->Move(tailx_, taily_, xdimd[dir], ydimd[dir]);
        dir = NUM_DIRECTIONS;
      }
    }
  }

  return result;
}


src/main.cc
//
// Grosso: main.cc
//

// 2002/06/09 10:29:50 tommi

///////////////////////////////////////////////////////////////////////////

#ifndef MAKEDEPEND
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <pwd.h>
#include <sys/types.h>
#endif

#include "config.hh"
#include "ScoreBoard.hh"
#include "Game.hh"
#include "version.hh"

///////////////////////////////////////////////////////////////////////////

static void
usage() {
  fprintf(stderr,
	  "Usage: grosso [-v] [-f<num #'s>] [-s<sleeptime ms>] "
	  "[-h<height>] [-w<width>]\n");
}

///////////////////////////////////////////////////////////////////////////

int
main(int argc, char* argv[])
{
  int i;
  int numfreds = 3;
  int sleeptime = 120;
  bool version = false;
  int width = 0;
  int height = 0;
  
  for (i = 1; i < argc; i++) {
    if (argv[i][0] == '-') {
      switch (argv[i][1]) {
      case 'f':
	numfreds = atoi(argv[i]+2);
	break;
      case 'h':
	height = atoi(argv[i]+2);
	break;
      case 's':
	sleeptime = atoi(argv[i]+2);
	break;
      case 'v':
	version = true;
	break;
      case 'w':
	width = atoi(argv[i]+2);
	break;
      default:
	usage();
	return 1;
      }
    }
    else {
      // all paramters without the leading dash are erroneous
      usage();
      return 1;
    }
  }
  
  if (version) {
    fprintf(stderr, "This is Grosso version " VERSION "\n");
  }
  else {
    srand(time(0));
    int score = Game::Play(width, height, numfreds, sleeptime);
    ScoreBoard sb(SCOREFILE);
    int rank = sb.Add(score, getpwuid(getuid())->pw_name);
    sb.Show(stdout, rank, score);
  }
  
  return 0;
}


src/Field.cc
//
// Grosso: Field.cc
//

// 2002/06/09 10:29:50 tommi

///////////////////////////////////////////////////////////////////////////

#ifndef MAKEDEPEND
#include <unistd.h>
#include <string.h>
#endif

#include "config.hh"
#include "Field.hh"

const int BUFSIZE = 1024;

using namespace std;

///////////////////////////////////////////////////////////////////////////

Field::Field(Dim width, Dim height, ColourNum fg, ColourNum bg) : nColours_(0)
{
  int i;
  char s[BUFSIZE];

  // curses and screen initialization
  initscr();
  cbreak();
  noecho();
  nonl();
  curs_set(0);
  if (has_colors()) {
    start_color();
  }
  fieldColourNum_ = NewColour(fg, bg);

  // get game, window and screen size
  getmaxyx(stdscr, screenHeight_, screenWidth_);
  if (width > 15) {
    width_ = MIN(screenWidth_, width);
  }
  else {
    width_ = screenWidth_;
  }
  if (height > 15) {
    height_ = MIN(screenHeight_, height);
  }
  else {
    height_ = screenHeight_;
  }

  //
  wIndent_ = 0;
  if (screenWidth_ > width_) {
    wIndent_ = (screenWidth_ - width_) / 2;
  }
  hIndent_ = 0;
  if (screenHeight_ > height_) {
    hIndent_ = (screenHeight_ - height_) / 2;
  }

#ifdef AVOID_MVWINCH
  aargh.reserve(width_*height_);
#endif // AVOID_MVWINCH

  // create and initialize game window
  window_ = newwin(height_, width_, hIndent_, wIndent_);
  wbkgdset(window_, COLOR_PAIR(1));
  nodelay(window_, TRUE);
  leaveok(window_, TRUE);
  wclear(window_);

  // set up dimension lookup tables for moving
  xdim_.push_back(width_-1);
  for (i = 0; i < width_; i++) {
    xdim_.push_back(i);
  }
  xdim_.push_back(0);
  ydim_.push_back(height_-1);
  for (i = 0; i < height_; i++) {
    ydim_.push_back(i);
  }
  ydim_.push_back(0);

  // count levels
  for (numLevels_ = 0;
       sprintf(s, "%s/%02d", LEVELDIR, numLevels_+1), access(s, R_OK) == 0;
       numLevels_++)
    ; // Empty
}

///////////////////////////////////////////////////////////////////////////

Field::~Field()
{
  // delete game window
  delwin(window_);
  // shut down curses
  endwin();
}

///////////////////////////////////////////////////////////////////////////

void
Field::Clear()
{
  wclear(window_);

#ifdef AVOID_MVWINCH
  for (int i = 0; i < width_*height_; i++) {
    aargh[i] = EMPTY;
  }
#endif // AVOID_MVWINCH

  touchwin(stdscr);
  wrefresh(stdscr);
}

///////////////////////////////////////////////////////////////////////////

void
Field::Puts(Dim x, Dim y, const string& s, ColourNum c)
{
  if (c == -1) {
    c = fieldColourNum_;
  }

  AttrOn(COLOR_PAIR(c));

#ifdef AVOID_MVWINCH
  for (unsigned i = 0; i < s.size(); i++) {
    if (s[i] == '\n') {
      y++;
      x = 0;
    }
    else{
      PutItem(x, y, s[i], c);
      x++;
    }
  }
#else
  mvwaddstr(window_, y, x, s.c_str());
#endif // AVOID_MVWINCH

  AttrOff(COLOR_PAIR(c));
}

///////////////////////////////////////////////////////////////////////////

int
Field::Display(const string& message, ColourNum c, Dim dx, Dim dy)
{
  int key;
  WINDOW *w;
  Dim x, y;
  char m[1024];
  char *line;
  short i;

  // show the message in a temporary window
  w = newwin(height_, width_, hIndent_, wIndent_);
  wbkgdset(w, COLOR_PAIR(1));
  wattron(w, COLOR_PAIR(c));
  wclear(w);
  x = (width_ - dx) / 2;
  y = (height_ - dy) / 2;
  strcpy(m, message.c_str());
  for (line = strtok(m, "\n"), i = 0;
       line != NULL;
       line = strtok(NULL, "\n"), ++i)
  {
    mvwaddstr(w, y+i, x, line);
  }
  wrefresh(w);
  while ((key = Key()) == ERR) {
    Sleep(100);
  }
  wattroff(w, COLOR_PAIR(c));
  delwin(w);

  touchwin(stdscr);
  wrefresh(stdscr);
  touchwin(window_);
  wrefresh(window_);

  return key;
}

///////////////////////////////////////////////////////////////////////////

void
Field::DrawLevel(FILE* f, Dim x, Dim y, Dim width, Dim height, ColourNum c)
{
  int i, j;
  char s[BUFSIZE];

  // draw the level into the specified position
  for (j = 0; j < height && fgets(s, sizeof(s), f); j++) {
    for (i = 0; i < width && s[i] != '\n'; i++) {
      if (s[i] != ' ') {
        PutItem(x + i, y + j, 'X', c);
      }
    }
  }
}

///////////////////////////////////////////////////////////////////////////

void
Field::Level(int level)
{
  char s[BUFSIZE];
  Dim x, y, w, h;
  FILE* lf;

  Clear();
  sprintf(s, "%s/%02d", LEVELDIR, (level - 1) % numLevels_ + 1);
  lf = fopen(s, "r");
  if (lf) {
    AttrOn(Field::REVERSE);
    // discard comment and get level size
    if (fgets(s, sizeof(s), lf) != NULL && fscanf(lf,"%d %d\n",&w,&h) > 0) {
      for (y = 0; y < Height(); y += h + 3) {
        for (x = 0; x < Width(); x += w + 3) {
          rewind(lf);
          // discard comment and dimensions lines
          if (fgets(s,sizeof(s),lf) != NULL && fgets(s,sizeof(s),lf) !=NULL) {
            DrawLevel(lf, x, y, MIN(Width()-x, w), MIN(Height()-y, h), fieldColourNum_);
          }
        }
      }
    }
    AttrOff(Field::REVERSE);
    fclose(lf);
  }
}

///////////////////////////////////////////////////////////////////////////

Field::ColourNum
Field::NewColour(ColourNum fg, ColourNum bg)
{
  init_pair(++nColours_, fg, bg);
  return nColours_;
}


src/ScoreBoard.cc
//
// Grosso: ScoreBoard.cc
//

// 2002/06/09 10:29:50 tommi

///////////////////////////////////////////////////////////////////////////

#ifndef MAKEDEPEND
//#include <iomanip.h>
#include <time.h>
#endif

#include "config.hh"
#include "ScoreBoard.hh"

using namespace std;

///////////////////////////////////////////////////////////////////////////

ScoreBoard::ScoreBoard(const string& scoreFile, int numScores) :
  numScores_(numScores)
{
  scoreFile_ = scoreFile;
  FILE *sf = fopen(scoreFile_.c_str(), "r");

  if (sf == NULL) {
    fprintf(stderr, "Couldn't open high score file in read mode.\n");
    perror(scoreFile_.c_str());
  }
  else {
    for (int i = 0; i < numScores_ && !feof(sf); i++) {
      char s[1024];
      if (fgets(s, 1024, sf)) {
        scores_.push_back(s);
      }
    }

    fclose(sf);
  }
}

///////////////////////////////////////////////////////////////////////////

ScoreBoard::~ScoreBoard()
{
  FILE *sf = fopen(scoreFile_.c_str(), "w");

  if (sf == NULL) {
    fprintf(stderr, "Couldn't open high score file in write mode.\n");
    perror(scoreFile_.c_str());
  }
  else {
    while(!scores_.empty()) {
      fprintf(sf, "%s", scores_.front().c_str());
      scores_.pop_front();
    }

    fclose(sf);
  }
}

///////////////////////////////////////////////////////////////////////////

int
ScoreBoard::Add(int score, const string& name)
{
  char s[256];
  time_t t;

  t = time(0);
  sprintf(s, "%6d  %10.10s %4.4s  %s\n",
	  score, ctime(&t), ctime(&t)+20, name.c_str());

  const ScoreEntry entry = s;
  Scores::iterator si;
  int i;
  for (i = 1, si = scores_.begin();
       si != scores_.end() && entry < *si;
       si++, i++)
    ;

  if (i <= numScores_) {
    scores_.insert(si, entry);
  }

  return i;
}

///////////////////////////////////////////////////////////////////////////

void
ScoreBoard::Show(FILE* f, int rank, int score)
{
  int i;
  Scores::const_iterator si;

  for (i = 1, si = scores_.begin(); si != scores_.end(); si++, i++) {
    fprintf(f, "%s %2d. %s", rank==i ? "O ":"  ", i, si->c_str());
  }

  if (rank >= i) {
    fprintf(f, "  You scored %d\n", score);
  }
}


src/Game.cc
//
// Grosso: Game.cc
//

// 2004/08/06 12:25:25 tommi

///////////////////////////////////////////////////////////////////////////

#ifndef MAKEDEPEND
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#endif

#include "config.hh"
#include "Game.hh"

using namespace std;

///////////////////////////////////////////////////////////////////////////

Game::Game(int width, int height, int numFreds, int sleepTime)
{
  // make sure installed correctly
  if (access(LEVELDIR, R_OK) != 0) {
    //cerr << "Can't access Grosso level directory \"" LEVELDIR "\"" << endl;
    fprintf(stderr, "Can't access Grosso level directory \"" LEVELDIR "\"\n");
    sleepTime_ = -1;
    return;
  }

  // initialize field and grosso
  field_ = new Field(width, height, COLOR_BLUE, COLOR_WHITE);
  foodColourNum_ = field_->NewColour(COLOR_CYAN, COLOR_WHITE);
  textColourNum_ = field_->NewColour(COLOR_BLACK, COLOR_WHITE);
  grosso_ = new Grosso(field_, 'O', COLOR_BLACK, COLOR_WHITE);

  // create freds
  for (int i = 0; i < numFreds; i++) {
    Fred f(field_, '#', COLOR_RED, COLOR_WHITE, 3);
    freds_.push_back(f);
  }

  sleepTime_ = sleepTime;
  foodTimer_ = -999;
}

///////////////////////////////////////////////////////////////////////////

Game::~Game()
{
  delete grosso_;
  delete field_;
}

///////////////////////////////////////////////////////////////////////////

int
Game::Pause(const string& shape)
{
  int i = 0;
  int key;

  field_->Flush();
  field_->Refresh();
  do {
    // cycle grosso's head shape
    if (shape.size() > 1) {
      grosso_->ShowHead(shape[i % shape.size()]);
      field_->Refresh();
    }

    key = field_->Key();
    if (key == '?' || key == 'h') {
      Help();
      key = ERR;
    }
    else if (key == 'q') {
      key = ConfirmQuit();
      if (key == 'y') {
	EndGame();
      }
      else {
	key = ERR;
      }
    }
    else {
      Sleep(sleepTime_ / ((shape.size() > 1) ? shape.size() : 1));
    }
    i++;
  } while (key == ERR || (shape.size() > 1 && i < 42/2));

  return key;
}

///////////////////////////////////////////////////////////////////////////

int
Game::Play(int width, int height, int numFreds, int sleepTime)
{
  int score;
  Game game(width, height, numFreds, sleepTime);

  if (game.sleepTime_ != -1) {
    score = game.Play();
  }
  else {
    score = -1;
  }

  return score;
}

///////////////////////////////////////////////////////////////////////////

int
Game::Play()
{
  score_ = 0;
  int level = 1;
  for(nGrossoes_ = 3; nGrossoes_ > 0; /* Empty */) {
    if (PlayLevel(level)) {
      level++;
    }
    else {
      nGrossoes_--;
    }
  }

  return score_;
}

///////////////////////////////////////////////////////////////////////////

bool
Game::PlayLevel(int level)
{
  Freds::iterator fi;
  Field::Dim x, y;

  foodOnField_ = false;
  numFoods_ = 10;
  field_->Level(level);
  grosso_->Reset(0, 0, 2, 10);
  InitFrame();
  for (fi = freds_.begin(); fi != freds_.end(); fi++) {
    do {
      x = rand()%field_->Width();
      y = rand()%field_->Height();
    } while ((x < 5 && y < 5) || field_->GetItem(x, y) != Field::EMPTY);
    fi->Reset(x, y, rand(), 10);
  }
  field_->Refresh();

  running_ = true;
  Pause();
  while (running_) {
    Sleep(sleepTime_);
    FoodOnField();
    MoveFreds();
    if (MoveGrosso(ProcessKeyPress()) == Grosso::ATE) {
      numFoods_--;
      if (numFoods_ == 0) {
        InitFrame();
        running_ = false;
        Pause("Oo.o");
      }
    }
    InitFrame();
  }

  return numFoods_ == 0;
}

///////////////////////////////////////////////////////////////////////////

void
Game::InitFrame()
{
  char s[16];

  if (foodTimer_ > 0) {
    foodTimer_--;
  }

  sprintf(s, "F%02d-%06d-G%02d", numFoods_, score_, nGrossoes_);
  field_->Puts((field_->Width()-strlen(s))/2, 0, s);

  field_->Refresh();
}

///////////////////////////////////////////////////////////////////////////

void
Game::FoodOnField()
{
  Field::Dim x, y;

  // if not already dropped, try to drop food into the field
  if (!foodOnField_) {
    if (rand()%100 < 10) {
      x = rand()%field_->Width();
      y = rand()%field_->Height();
      if (field_->GetItem(x, y) == Field::EMPTY) {
        field_->PutItem(x, y, Worm::FOOD, foodColourNum_);
        foodOnField_ = true;
        foodTimer_ = field_->Width() + field_->Height();
      }
    }
  }
}

///////////////////////////////////////////////////////////////////////////

int
Game::ProcessKeyPress()
{
  int key;
  int turn = Grosso::FORWARD;

  switch(key = field_->Key()) {
  case 'q':
  case ERR:
    break;
  case 'v':
  case '4':
    turn = Grosso::LEFT;
    break;
  case 'n':
  case '6':
    turn = Grosso::RIGHT;
    break;
  case '?':
  case 'h':
    Help();
    key = Pause();
    break;
  default:
    key = Pause();
    break;
  }
  if (key == 'q') {
    if (ConfirmQuit() == 'y') {
      EndGame();
    }
    else {
      Pause();
    }
  }

  return turn;
}

///////////////////////////////////////////////////////////////////////////

Grosso::MoveResult
Game::MoveGrosso(int turn)
{
  Grosso::MoveResult result;

  switch (result = grosso_->TryMove(turn)) {
  case Worm::ATE:
    score_ += 10 + foodTimer_;
    foodOnField_ = false;
    break;
  case Worm::CRASHED:
    running_ = false;
    Pause("\\|/-");
  default:
    break;
  }

  return result;
}

///////////////////////////////////////////////////////////////////////////

void
Game::MoveFreds()
{
  Freds::iterator fi;

  for (fi = freds_.begin(); fi != freds_.end(); fi++) {
    switch (fi->TryMove()) {
    case Worm::ATE:
      score_ += 10 + foodTimer_;
      foodOnField_ = false;
      break;
    case Worm::GOT_BLOCKED:
      score_++;
      break;
    case Worm::DIED:
      score_ += 100;;
      break;
    default:
      break;
    }
  }
}

///////////////////////////////////////////////////////////////////////////

void
Game::EndGame()
{
  running_ = false;
  nGrossoes_ = 0;
}

///////////////////////////////////////////////////////////////////////////

int
Game::ConfirmQuit()
{
  return field_->Display
    ("Are you sure you want to quit Grosso?"
     " \n \n"
     "            Press y or n.",
     textColourNum_, 37, 3);
}

///////////////////////////////////////////////////////////////////////////

void
Game::Help()
{
  field_->Display
    ("    *** This is Grosso, a simple character terminal game ***\n"
     " \n"
     "  You control the O-headed worm. Don't crash into anything,\n"
     "  eat all *'s, and you are on the next level! Be careful with and\n"
     "  exploit the wraparound. See the man page for more info.\n"
     " \n"
     "  Keys: v = turn left\n"
     "        n = turn right\n"
     "    space = pause\n"
     "        h = this help\n"
     "        q = quit\n"
     " \n"
     "  This program is PostcardWare: As you have come as far as reading\n"
     "  this message, you might as well send a picture postcard to: *B->\n"
     " \n"
     "    Tommi Joutsiniemi\n"
     "    Lukinkatu 2d\n"
     "    33580 Tampere\n"
     "    Finland\n"
     " \n"
     "  Press any key to resume game.\n",
     textColourNum_, 66, 21);
}
